presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1695,6 +1695,18 @@ protected Scope visitExcept(Except node, Optional<Scope> scope)
             return visitSetOperation(node, scope);
         }
 
+        private <T extends Expression> List<Expression> extractDereferenceExpressionsAndIdentifiers(T expression)
+        {
+            if (expression instanceof DereferenceExpression || expression instanceof Identifier) {
+                return ImmutableList.of(expression);
+            }
+            ImmutableList.Builder<Expression> dereferenceExpression = new ImmutableList.Builder<>();
```

Comment:
Also it might be better to use ImmutableSet and just send the builder around as an argument will need a helper in that case that does the actual recursion.

Reply:
Also, how does this work with struct fields? ON t1.a.b.c = T2.x.y.z should be fine. Add a test for that as well.